core/src/ballot_box.rs

```diff
@@ -0,0 +1,378 @@
+
+    operators_voted: PodU64,
+    unique_ballots: PodU64,
+
+    //TODO fix 32 -> MAX_OPERATORS
+    operator_votes: [OperatorVote; 32],
+    ballot_tallies: [BallotTally; 32],
+}
+
+impl Discriminator for BallotBox {
+    const DISCRIMINATOR: u8 = Discriminators::EpochSnapshot as u8;
+}
+
+impl BallotBox {
+    pub fn new(ncn: Pubkey, ncn_epoch: u64, bump: u8, current_slot: u64) -> Self {
+        Self {
+            ncn,
+            ncn_epoch: PodU64::from(ncn_epoch),
+            bump,
+            slot_created: PodU64::from(current_slot),
+            slot_consensus_reached: PodU64::from(0),
+            operators_voted: PodU64::from(0),
+            unique_ballots: PodU64::from(0),
+            //TODO fix 32 -> MAX_OPERATORS
+            operator_votes: [OperatorVote::default(); 32],
+            ballot_tallies: [BallotTally::default(); 32],
+            reserved: [0; 128],
+        }
+    }
+
+    pub fn seeds(ncn: &Pubkey, ncn_epoch: u64) -> Vec<Vec<u8>> {
+        Vec::from_iter(
+            [
+                b"ballot_box".to_vec(),
+                ncn.to_bytes().to_vec(),
+                ncn_epoch.to_le_bytes().to_vec(),
+            ]
+            .iter()
+            .cloned(),
+        )
+    }
+
+    pub fn find_program_address(
+        program_id: &Pubkey,
+        ncn: &Pubkey,
+        ncn_epoch: u64,
+    ) -> (Pubkey, u8, Vec<Vec<u8>>) {
+        let seeds = Self::seeds(ncn, ncn_epoch);
+        let seeds_iter: Vec<_> = seeds.iter().map(|s| s.as_slice()).collect();
+        let (pda, bump) = Pubkey::find_program_address(&seeds_iter, program_id);
+        (pda, bump, seeds)
+    }
+
+    pub fn load(
+        program_id: &Pubkey,
+        ncn: &Pubkey,
+        ncn_epoch: u64,
+        epoch_snapshot: &AccountInfo,
+        expect_writable: bool,
+    ) -> Result<(), ProgramError> {
+        if epoch_snapshot.owner.ne(program_id) {
+            msg!("Ballot box account has an invalid owner");
+            return Err(ProgramError::InvalidAccountOwner);
+        }
+        if epoch_snapshot.data_is_empty() {
+            msg!("Ballot box account data is empty");
+            return Err(ProgramError::InvalidAccountData);
+        }
+        if expect_writable && !epoch_snapshot.is_writable {
+            msg!("Ballot box account is not writable");
+            return Err(ProgramError::InvalidAccountData);
+        }
+        if epoch_snapshot.data.borrow()[0].ne(&Self::DISCRIMINATOR) {
+            msg!("Ballot box account discriminator is invalid");
+            return Err(ProgramError::InvalidAccountData);
+        }
+        if epoch_snapshot
+            .key
+            .ne(&Self::find_program_address(program_id, ncn, ncn_epoch).0)
+        {
+            msg!("Ballot box account is not at the correct PDA");
+            return Err(ProgramError::InvalidAccountData);
+        }
+        Ok(())
+    }
+
+    fn slot_consensus_reached(&self) -> u64 {
+        self.slot_consensus_reached.into()
+    }
```

Comment:
So you can neatly interface with this account for rewards processing:
would add a 

```
fn is_consensus_reached() -> bool {
  self.slot_consensus_reached() == 0
}
```
as well as a way to quickly get the winning ballot, either returning an `Option<Ballot>` from tally_votes, or including `winning_ballot: Option<Ballot>` on the struct

Reply:
Done